### PR TITLE
[TensorIR][M2a] Reduction Factoring (RFactor)

### PR DESCRIPTION
**Pull request by MasterJH5574**
This PR is part of the TensorIR upstreaming effort (#7527), which adds the one schedule primitive "rfactor".

cc @junrushao1994 @tqchen @comaniac @jcf94

### CODE REVIEW

File: include/tvm/tir/schedule/block_scope.h [hunk 1]
```diff
@@ -255,14 +255,14 @@ class BlockScopeNode : public Object {
 class BlockScope : public ObjectRef {
  public:
   /*! \brief The constructor creating an empty block scope with on dependency information */
-  TVM_DLL BlockScope();
+  TVM_DLL explicit BlockScope();
```
Comment (junrushao1994) [inline include/tvm/tir/schedule/block_scope.h:258]: No explicit is needed on empty constructor

File: include/tvm/tir/schedule/schedule.h [hunk 2]
```diff
@@ -242,6 +242,17 @@ class ScheduleNode : public runtime::Object {
   /******** Schedule: loop binding/annotation ********/
   /******** Schedule: cache read/write ********/
   /******** Schedule: reduction ********/
+  /*!
+   * \brief Factor a reduction block by the specified loop
+   * \details See python/tvm/tir/schedule/schedule.py
```
Comment (junrushao1994) [inline include/tvm/tir/schedule/schedule.h:247]: Hmmmm I am not sure, but perhaps it is better to have one complete doc on C++ side, another on python side, so probably you might want to copy from python's doc

File: include/tvm/tir/stmt.h [hunk 3]
```diff
@@ -1361,6 +1362,23 @@ TVM_DLL PrimExpr TypeAnnotation(DataType dtype, Span span = Span());
 // overload printing of for type.
 TVM_DLL std::ostream& operator<<(std::ostream& os, ForKind kind);
 
+// inline implementations
+inline const char* IterVarType2String(ForKind t) {
```
Comment (junrushao1994) [inline include/tvm/tir/stmt.h:1366]:
```suggestion
inline const char* ForKind2String(ForKind t) {
```

File: src/tir/ir/expr.cc [hunk 4]
```diff
@@ -21,13 +21,10 @@
  * \file expr.cc
  */
 #include <tvm/runtime/registry.h>
-#include <tvm/tir/expr.h>
+#include <tvm/tir/analysis.h>
```
Comment (junrushao1994) [inline src/tir/ir/expr.cc:24]: Recover analysis.h back to expr.h

File: src/tir/schedule/state.cc [hunk 5]
```diff
@@ -737,7 +709,8 @@ class SRefUpdater : public StmtVisitor {
   void UpdateBlockInfo(const StmtSRef& block_sref) {
     using TIter = std::unordered_map<StmtSRef, BlockInfo, ObjectPtrHash, ObjectPtrEqual>::iterator;
     // The caller is responsible for correcting the flags
-    BlockInfo new_info(BlockScope(GetChildBlocks(self_, block_sref)));
+    Array<StmtSRef> child_block_srefs = GetChildBlockSRefOnSRefTree(self_, block_sref);
+    BlockInfo new_info((BlockScope(child_block_srefs)));
```
Comment (junrushao1994) [inline src/tir/schedule/state.cc:713]:
```suggestion
    BlockInfo new_info((BlockScope(GetChildBlockSRefOnSRefTree(self_, block_sref))));
```

File: include/tvm/tir/schedule/schedule.h (hunk 2 above)
Comment (MasterJH5574) [reply include/tvm/tir/schedule/schedule.h:247]: I don't know how to present the code examples from Python side here. Do you think it's acceptable to omit the code examples?

Comment (junrushao1994) [reply include/tvm/tir/schedule/schedule.h:247]: We probably dont need the code examples, because it is python-specific

Comment (MasterJH5574) [reply include/tvm/tir/schedule/schedule.h:247]: Got it. Will do that tomorrow.

File: tests/python/unittest/test_tir_schedule_reduction.py [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+# pylint: disable=missing-function-docstring,missing-module-docstring
+import pytest
+
+import numpy as np
+import tvm
+import tvm.testing
+from tvm import tir
+from tvm.script import ty
+
+# pylint: disable=no-member,invalid-name,unused-variable
```
Comment (comaniac) [inline tests/python/unittest/test_tir_schedule_reduction.py:26]: Merge to L17 or move to L18 if it's too long

**Review by comaniac — COMMENTED** [review]
Leave the core rfactor implement to tomorrow.
Also cc @jcf94 @vinx13 to help review.

File: python/tvm/tir/schedule/schedule.py [hunk 7]
```diff
@@ -512,6 +512,150 @@ def after_inline(a: ty.handle, c: ty.handle) -> None:
...
+            print(tvm.script.asscript(sch.mod["main"]))
+
+        After applying rfactor, the IR becomes:
+
+        .. code-block:: python
+
+            @tvm.script.tir
+            def after_rfactor(a: ty.handle, b: ty.handle) -> None:
+                A = tir.match_buffer(a, [128, 128, 128])
+                B = tir.match_buffer(b, [128])
```
[the remark this answers is not in the document]
Comment (MasterJH5574) [reply python/tvm/tir/schedule/schedule.py:623]: Oh thanks! I slipped through this line. Usually we just leave out the type specification in `tir.match_buffer`, so I'll remove those in the "before".

File: python/tvm/tir/schedule/schedule.py [hunk 8]
```diff
@@ -512,6 +512,150 @@ def after_inline(a: ty.handle, c: ty.handle) -> None:
...
+        1) `loop` has only one child block, and it is a reduction block;
+        2) `loop` is a reduction loop, i.e. the loop variable is bound to only reduction variables
+        in the block binding;
+        3) `loop` is not parallelized, vectorized, unrolled or bound to any thread axis;
+        4) The block scope that `loop` is in is a staged-pipeline;
+        5) The outermost loop outside the reduction block should has the reduction block as its first child block;
+        6) The outermost reduction loop should have only one child block;
+        7) An unary extent loop that is not bound to any reduction or data parallel variables in the block binding
+        should not appear under some reduction loop;
+        8) The reduction block should write to only one buffer, and its init and body block only is
```
[the remark this answers is not in the document]
Comment (MasterJH5574) [reply python/tvm/tir/schedule/schedule.py:648]: Yeah I believe that was a mistake. I'll rephrase it.

**Review by comaniac — APPROVED** [review]
LGTM. We should be able to merge this PR when CI is green.

**Comment by tqchen** [comment]
@MasterJH5574 , please fix the docstring issues. You can reproduce by following ./tests/scripts/task_sphinx_precheck.sh

**Review by jcf94 — APPROVED** [review]

**Comment by junrushao1994** [comment]
@tqchen looks like this PR is experiencing CPU time out on unittests

**Review by junrushao1994 — APPROVED** [review]

**Pull request merged**